divide_and_conquer/mergesort.py

```diff
@@ -1,48 +1,58 @@
-def merge(arr, left, mid, right):
-    # overall array will divided into 2 array
-    # left_arr contains the left portion of array from left to mid
-    # right_arr contains the right portion of array from mid + 1 to right
-    left_arr = arr[left : mid + 1]
-    right_arr = arr[mid + 1 : right + 1]
-    k = left
-    i = 0
-    j = 0
-    while i < len(left_arr) and j < len(right_arr):
-        # change sign for Descending order
-        if left_arr[i] < right_arr[j]:
-            arr[k] = left_arr[i]
-            i += 1
+def merge(left_half, right_half):
+    sorted_array = [None] * (len(right_half) + len(left_half))
+
+    pointer1 = 0  # pointer to current index for left Half
+    pointer2 = 0  # pointer to current index for the right Half
+    index = 0  # pointer to current index for the sorted array Half
+
+    while pointer1 < len(left_half) and pointer2 < len(right_half):
+        if left_half[pointer1] < right_half[pointer2]:
+            sorted_array[index] = left_half[pointer1]
+            pointer1 += 1
+            index += 1
         else:
-            arr[k] = right_arr[j]
-            j += 1
-        k += 1
-    while i < len(left_arr):
-        arr[k] = left_arr[i]
-        i += 1
-        k += 1
-    while j < len(right_arr):
-        arr[k] = right_arr[j]
-        j += 1
-        k += 1
-    return arr
-
-
-def mergesort(arr, left, right):
+            sorted_array[index] = right_half[pointer2]
+            pointer2 += 1
+            index += 1
+    while pointer1 < len(left_half):
+        sorted_array[index] = left_half[pointer1]
+        pointer1 += 1
+        index += 1
+
+    while pointer2 < len(right_half):
+        sorted_array[index] = right_half[pointer2]
+        pointer2 += 1
+        index += 1
+
+    return sorted_array
+
+
+def merge_sort(array):
     """
-    >>> mergesort([3, 2, 1], 0, 2)
-    [1, 2, 3]
-    >>> mergesort([3, 2, 1, 0, 1, 2, 3, 5, 4], 0, 8)
-    [0, 1, 1, 2, 2, 3, 3, 4, 5]
+    >>> from random import shuffle
+    >>> array = [-2, 3, -10, 11, 99, 100000, 100, -200]
+    >>> shuffle(array)
+    >>> merge_sort(array)
+    [-200, -10, -2, 3, 11, 99, 100, 100000]
+    >>> shuffle(array)
+    >>> merge_sort(array)
+    [-200, -10, -2, 3, 11, 99, 100, 100000]
+    >>> shuffle(array)
+    >>> merge_sort(array)
+    [-200, -10, -2, 3, 11, 99, 100, 100000]
```

Comment:
Please use multiple test cases, some of the examples could be:
- Empty list
- One element list
- List containing strings, integers and other lists
- Mixed element types like a list containing integers, strings, etc
- Use `from string import letters`, `range`
- Verify the tests with built-in `sort` method or `sorted` function (easier than typing it out)

Be creative and mix it up in coming up with the test cases 😁